p4-16/spec/P4-16-spec.mdk

```diff
@@ -3605,6 +3605,44 @@ tuple<bit<32>, bool> x = { 10, false };
 
 The empty list expression has type `tuple<>` - a tuple with no components.
 
+## Structure-valued expressions { #sec-structure-expressions}
+
+One can write expressions that evaluate to a structure or header.  The
+syntax of these expressions is given by:
+
+~ Begin P4Grammar
+expression ...
+    | '{' kvList '}'
+    | typeName '{' kvList '}'
```

Comment:
This creates the awkward situation where all of the following are valid syntax:
* `{ a=0, b=1 }`
* `S { a=0, b=1 }`
* `(S) { a=0, b=1 }`
* `(S1) S2 { a=0, b=1 }`

One way around this awkwardness would be to only allow the `'{' kvList '}'` production and remove the `typeName '{' kvList '}'` production, but then typing of struct-valued expressions can become ambiguous when the program has defined multiple types that match.

Another alternative would be demote `'{' kvList '}'` from being a first-class expression, and only allow it as a struct initializer: `S s = { a=0, b=1 };`. Then, struct-valued expressions would always use the `S { a=0, b=1 }` form. My impression is that this would be compatible with what we released in 1.2.0, but I haven't confirmed.

Reply:
the typechecker will reject today casts with struct types.

Reply:
maybe that's a good rule: if a cast is to a Struct or header type then it is only allowed for struct-valued expressions.